lib/tree.rb

```diff
@@ -1,66 +1,235 @@
-  # Space Complexity: 
+  
+  # IN CLASS PRACTICE: inorder() recursive
+  # def inorder_helper(current_node, list = [])
+  #   return list if !current_node 
+  #   inorder_helper(current_node.left, list)
+  #   list << {key: current_node.key, value: current_node.value}
+  #   inorder_helper(current_node.right, list)
+  #   return list
+  # end
+  
+  # def inorder
+  #   return inorder_helper(@root)
+  # end
+  
+  # Time Complexity: O(n) with n is the number of nodes in the tree
+  # Space Complexity: O(n)
+  def preorder_helper(current_node, list = [])
+    return list if !current_node
+    list << {key: current_node.key, value: current_node.value}
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+    return list
+  end
+  
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(n) with n is the number of nodes in the tree
+  # Space Complexity: O(n)
+  def postorder_helper(current_node, list = [])
+    return list if !current_node
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list << {key: current_node.key, value: current_node.value}
+    return list
+  end
+  
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(n) with n is the number of nodes in the tree
+  # Space Complexity: O(logn) which is also the height of the tree.
+  def height_helper(current_node, level = 0)
+    return level if !current_node
+    level = [height_helper(current_node.left, level + 1), height_helper(current_node.right, level + 1)].max
+    return level
+  end
+  
   def height
-    raise NotImplementedError
+    return height_helper(@root)
   end
-
+  
   # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def bfs
-    raise NotImplementedError
+    list = []
+    if @root
+      queue = []
+      queue << @root
+      queue.each do |current|
+        list << {key: current.key, value: current.value}
+        queue << current.left if current.left
+        queue << current.right if current.right
+      end
+    end
+    return list
   end
+  
+  def delete(key)
+    current_and_parent_pair = find_current_and_parent_nodes(key)
+    current = current_and_parent_pair[:current] 
+    parent = current_and_parent_pair[:parent]
+    if current 
+      remove_child(parent, current)
+      new_subtree = nil   
+      # rearrange new subtree from children if current is not a leaf node
+      if current.left || current.right
+        left = current.left 
+        right = current.right 
+        right_subtree_leftmost = find_leftmost_node(right)
+        right_subtree_leftmost.left = left if right_subtree_leftmost
+        
+        new_subtree = right_subtree_leftmost ? right : left
```

Comment:
What is right is `nil` and left is not?

I don't think this completely works.  You have the right idea.

Reply:
If right is nil then find_leftmost_node will return nil. And line 171 will be skipped and on line 173 new_subtree will be left. And then the new_subtree which is left now will be linked to parent. I think this code covers that case too. I’ll add tests.

Reply:
@CheezItMan

Reply:
Ok that will work, thank you.